internal/sdk/cloudian/qos.go

```diff
@@ -0,0 +1,132 @@
+package cloudian
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+)
+
+// QualityOfService configures data limits for a Group or User.
+type QualityOfService struct {
+	// Soft is the soft (warning) limit.
+	Soft QualityOfServiceLimits
+	// Hard is the hard limit.
+	Hard QualityOfServiceLimits
+}
+
+// QualityOfService configures data limits.
+type QualityOfServiceLimits struct {
+	// StorageQuotaKBytes is the limit for total stored data in KiB.
+	StorageQuotaKBytes *int64
+	// StorageQuotaCount is the limit for total number of objects.
+	StorageQuotaCount *int64
+	// RequestsPerMin is the limit for number of HTTP requests per minute.
+	RequestsPerMin *int64
+	// InboundKBytesPerMin is the limit for inbound data per minute in KiB.
+	InboundKBytesPerMin *int64
+	// OutboundKBytesPerMin is the limit for outbound data per minute in KiB.
+	OutboundKBytesPerMin *int64
+}
+
+// CreateQuota sets the QoS limits for a `User`. To change QoS limits, a delete and recreate is necessary.
+func (client Client) CreateQuota(ctx context.Context, user User, qos QualityOfService) error {
+	rawParams := map[string]*int64{
+		"hlStorageQuotaKBytes": qos.Hard.StorageQuotaKBytes,
+		"wlStorageQuotaKBytes": qos.Soft.StorageQuotaKBytes,
+		"hlStorageQuotaCount":  qos.Hard.StorageQuotaCount,
+		"wlStorageQuotaCount":  qos.Soft.StorageQuotaCount,
+		"hlRequestRate":        qos.Hard.RequestsPerMin,
+		"wlRequestRate":        qos.Soft.RequestsPerMin,
+		"hlDataKBytesIn":       qos.Hard.InboundKBytesPerMin,
+		"wlDataKBytesIn":       qos.Soft.InboundKBytesPerMin,
+		"hlDataKBytesOut":      qos.Hard.OutboundKBytesPerMin,
+		"wlDataKBytesOut":      qos.Soft.OutboundKBytesPerMin,
+	}
+
+	params := make(map[string]string, len(rawParams))
+	for key, raw := range rawParams {
+		val := int64(-1)
+		if raw != nil {
+			val = *raw
+		}
+		if val < -1 {
+			val = -1
+		}
+		params[key] = strconv.FormatInt(val, 10)
```

Comment:
or `strconv.Itoa(int(val))`

Reply:
Wouldn't that downcast int64 to int32 on a 32 bit system?

Reply:
Exactly, `strconv.FormatInt(val, 10)` is the way to go here!